File: codechef/feb19b/xdcomp/Main.hs
import Data.Bits
import Data.Int
import Data.Graph
import Data.Tree
import Data.Array
import Control.Monad
import Data.Semigroup

ka :: Int64
ka = 10^9 + 7

mo :: Int64 -> Int64
mo x = x `mod` (10^9 + 7)

acc :: (Semigroup a) => Tree a -> Tree a
acc (Node a xs) = Node (foldl (<>) a [x | Node x _ <- xs']) xs' where
	xs' = map acc $ xs

rev :: (Semigroup a) => Tree a -> Tree a
rev (Node a xs) = Node (foldl (<>) a [x | Node x _ <- xs]) xs' where
	xs' = map rev $ xs

filt :: (a -> Bool) -> Tree a -> Forest a
filt p (Node n xs) = if p n then [Node n xs'] else xs' where
	xs' = xs >>= filt p

data Xor = Xor Int64
get :: Xor -> Int64
get (Xor a) = a

lift :: (Int64 -> Int64 -> Int64) -> Xor -> Xor -> Xor
lift f a b = Xor $ f (get a) (get b)

instance Semigroup Xor where
	(<>) = lift xor

instance Monoid Xor where
	mempty = Xor 0

process_zero :: Int -> [Int64] -> [(Int, Int)] -> Tree Int64
process_zero n weis ass = fmap (arr !) tree where
	arr = listArray bounds weis
	tree = case dff $ buildG bounds ass of
		[x] -> x
		_ -> error "Not a Tree"
	bounds = (0, n - 1)

process_one :: Int64 -> Tree Int64 -> Forest Bool
process_one i = (map $ fmap (==0))
	. (map $ fmap get)
	. (map $ rev)
	. filt (\(Xor x) -> (x == 0) || (x == i))
	. acc
	. fmap Xor

data Ways = Ways Int64 Int64
instance Semigroup Ways where
	(Ways a1 a2) <> (Ways b1 b2) = Ways (mo c1) (mo c2) where
		c1 = (mo $ a1 * b1) + (mo $ a1 * b2) + (mo $ a2 * b1)
		c2 = (mo $ a2 * b1) + (mo $ a2 * b2) + (mo $ a1 * b1)

make :: Bool -> Ways
make True = Ways 0 1
make False = Ways 1 0

process_two :: Tree Bool -> Int64
process_two t = x where
	Node (Ways x _) _ = acc $ fmap make t

process :: Int -> Int64 -> [Int64] -> [(Int, Int)] -> Int64
process n x weis ass = process_two tree where
	[tree] = process_one x all
	all = process_zero n weis ass

line_as :: (Read a) => IO [a]
line_as = do
	line <- getLine
	return $ map read $ words line


main :: IO ()
main = do
	[n', x] <- line_as :: IO [Int64]
	let n = fromIntegral n'
	weis <- line_as :: IO [Int64]

	ass' <- forM (take (n-1) $ repeat ())  (\_ -> do
		[u, v] <- line_as
		return [(u-1, v-1), (v-1, u-1)]) :: IO [[(Int, Int)]]
	let ass = join ass'
	let res = process n x weis ass
	putStrLn $ show res

File: codechef/feb19b/xdcomp/main.cpp
#include <bits/stdc++.h>
using namespace std;

typedef long long int ll;

const ll MOD = 1000000007;

struct Tree {
	vector<int> ch;
	int dad;
	ll val;
};

void acc(int root, vector<Tree> &tree) {
	for(int ch: tree[root].ch) {
		acc(ch, tree);
		tree[root].val ^= tree[ch].val;
	}
}

void rev(int root, vector<Tree> &tree) {
	for(int ch: tree[root].ch) {
		tree[root].val ^= tree[ch].val;
		rev(ch, tree);
	}
}

ll filter(ll x, int root, vector<Tree> &tree, vector<int> &v) {
	ll total = 0;
	if(tree[root].val == 0 || tree[root].val == x) {
		total++;
		vector<int> own;
		for(int ch : tree[root].ch) {
			total += filter(x, ch, tree, own);
		}
		tree[root].ch = own;
		v.push_back(root);
	} else {
		for(int ch : tree[root].ch) {
			total += filter(x, ch, tree, v);
		}
	}
	return total;
}

struct Ways {
	ll o;
	ll e;
};

Ways operator+(Ways one, Ways two) {
	ll o = 0;
	o += one.o * two.o % MOD;
	o %= MOD;
	o += one.o * two.e % MOD;
	o %= MOD;
	o += one.e * two.o % MOD;
	o %= MOD;

	ll e = 0;
	e += one.e * two.o % MOD;
	e %= MOD;
	e += one.e * two.e % MOD;
	e %= MOD;
	e += one.o * two.o % MOD;
	e %= MOD;

	return Ways{o, e};
}

Ways calc(int root, vector<Tree> &tree) {
	Ways curr;
	if(tree[root].val == 0) curr = Ways{0, 1};
	else curr = Ways{1, 0};

	for(int ch: tree[root].ch) {
		Ways rec = calc(ch, tree);
		curr = curr + rec;
	}

	return curr;
}

int main() {
	int n;
	cin >> n;
	ll x;
	cin >> x;

	vector<ll> weis(n);
	for(int i = 0; i < n; ++i) cin >> weis[i];

	vector<vector<int>> graph(n);
	for(int i = 1; i < n; ++i) {
		int u, v;
		cin >> u >> v;
		u--; v--;
		graph[u].push_back(v);
		graph[v].push_back(u);
	}

	vector<Tree> tree(n);

	function<void(int, int)> dfs = [&](int node, int dad) {
		tree[node].val = weis[node];
		for(int ch : graph[node]) if(ch != dad) {
			tree[node].ch.push_back(ch);
			dfs(ch, node);
		}
	};

	dfs(0, -1);

	acc(0, tree);
	vector<int> v;
	ll total = filter(x, 0, tree, v);

	if(v.size() != 1) {
		cout << "0" << endl;
		return 0;
	}

	if(x == 0) {
		ll res = 1;
		for(int i = 1; i < total; ++i) {
			res *= 2;
			res %= MOD;
		}

		cout << res << endl;
		return 0;
	}

	rev(0, tree);
	Ways w = calc(0, tree);
	cout << ((x == 0) ? w.e : w.o) << endl;
}
